File: packages/sdk-js/src/resources/bills.ts
import { HttpClient } from '../utils/request';
import { createPaginatedList } from '../utils/pagination';
import { PaginatedResponse, SingleResponse } from '../types/common.types';
import {
  Bill,
  BillListParams,
  BillVersion,
  BillDiff,
  BillDiffParams,
  Amendment,
} from '../types/bills.types';

export class BillsResource {
  private client: HttpClient;

  constructor(client: HttpClient) {
    this.client = client;
  }

  /**
   * List bills with pagination and filtering
   */
  async list(params: BillListParams = {}): Promise<PaginatedResponse<Bill>> {
    return this.client.get<PaginatedResponse<Bill>>('/v1/bills', {
      limit: params.limit,
      cursor: params.cursor,
      status: params.status,
      category: params.category,
      region: params.region,
      search: params.search,
    });
  }

  /**
   * Iterate through all bills matching the filters
   */
  listAll(params: Omit<BillListParams, 'cursor'> = {}) {
    const paginatedList = createPaginatedList<Bill, BillListParams>(
      (p) => this.list(p)
    );
    return paginatedList.listAll(params);
  }

  /**
   * Get a specific bill by ID
   */
  async get(id: string): Promise<Bill> {
    const response = await this.client.get<SingleResponse<Bill>>(`/v1/bills/${id}`);
    return response.data;
  }

  /**
   * Get version history for a bill
   */
  async getVersions(id: string): Promise<BillVersion[]> {
    const response = await this.client.get<SingleResponse<BillVersion[]>>(
      `/v1/bills/${id}/versions`
    );
    return response.data;
  }

  /**
   * Get diff between two versions of a bill
   */
  async diff(id: string, params: BillDiffParams = {}): Promise<BillDiff> {
    const response = await this.client.get<SingleResponse<BillDiff>>(
      `/v1/bills/${id}/diff`,
      {
        fromVersion: params.fromVersion,
        toVersion: params.toVersion,
      }
    );
    return response.data;
  }

  /**
   * Get amendments for a bill
   */
  async getAmendments(
    id: string,
    params: { limit?: number; cursor?: string } = {}
  ): Promise<PaginatedResponse<Amendment>> {
    return this.client.get<PaginatedResponse<Amendment>>(
      `/v1/bills/${id}/amendments`,
      params
    );
  }

  /**
   * Iterate through all amendments for a bill
   */
  listAllAmendments(id: string) {
    const paginatedList = createPaginatedList<Amendment, { limit?: number; cursor?: string }>(
      (p) => this.getAmendments(id, p)
    );
    return paginatedList.listAll({});
  }
}

File: packages/sdk-python/constitutional/resources/bills.py
"""Bills resource for the Constitutional SDK."""

from typing import Optional, Generator

from constitutional.utils.request import HttpClient
from constitutional.utils.pagination import paginate
from constitutional.types.common import PaginatedResponse, Pagination
from constitutional.types.bills import Bill, BillVersion, BillDiff, Amendment, BillStatus


class BillsResource:
    """Bills API resource."""

    def __init__(self, client: HttpClient):
        self._client = client

    def list(
        self,
        status: Optional[BillStatus] = None,
        category: Optional[str] = None,
        region: Optional[str] = None,
        search: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> PaginatedResponse[Bill]:
        """
        List bills with optional filtering.

        Args:
            status: Filter by bill status
            category: Filter by category
            region: Filter by region
            search: Search query
            limit: Maximum number of results
            cursor: Pagination cursor

        Returns:
            Paginated list of bills
        """
        response = self._client.get(
            "/v1/bills",
            params={
                "status": status,
                "category": category,
                "region": region,
                "search": search,
                "limit": limit,
                "cursor": cursor,
            },
        )

        return PaginatedResponse[Bill](
            data=[Bill(**b) for b in response["data"]],
            pagination=Pagination(**response["pagination"]),
        )

    def list_all(
        self,
        status: Optional[BillStatus] = None,
        category: Optional[str] = None,
        region: Optional[str] = None,
        search: Optional[str] = None,
    ) -> Generator[Bill, None, None]:
        """
        Iterate through all bills matching the filters.

        Args:
            status: Filter by bill status
            category: Filter by category
            region: Filter by region
            search: Search query

        Yields:
            Bills matching the filters
        """
        return paginate(
            self.list,
            status=status,
            category=category,
            region=region,
            search=search,
        )

    def get(self, bill_id: str) -> Bill:
        """
        Get a specific bill by ID.

        Args:
            bill_id: The bill ID

        Returns:
            The bill
        """
        response = self._client.get(f"/v1/bills/{bill_id}")
        return Bill(**response["data"])

    def get_versions(self, bill_id: str) -> list[BillVersion]:
        """
        Get version history for a bill.

        Args:
            bill_id: The bill ID

        Returns:
            List of bill versions
        """
        response = self._client.get(f"/v1/bills/{bill_id}/versions")
        return [BillVersion(**v) for v in response["data"]]

    def diff(
        self,
        bill_id: str,
        from_version: Optional[int] = None,
        to_version: Optional[int] = None,
    ) -> BillDiff:
        """
        Get diff between two versions of a bill.

        Args:
            bill_id: The bill ID
            from_version: Starting version
            to_version: Ending version

        Returns:
            Diff between versions
        """
        response = self._client.get(
            f"/v1/bills/{bill_id}/diff",
            params={
                "fromVersion": from_version,
                "toVersion": to_version,
            },
        )
        return BillDiff(**response["data"])

    def get_amendments(
        self,
        bill_id: str,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> PaginatedResponse[Amendment]:
        """
        Get amendments for a bill.

        Args:
            bill_id: The bill ID
            limit: Maximum number of results
            cursor: Pagination cursor

        Returns:
            Paginated list of amendments
        """
        response = self._client.get(
            f"/v1/bills/{bill_id}/amendments",
            params={"limit": limit, "cursor": cursor},
        )
        return PaginatedResponse[Amendment](
            data=[Amendment(**a) for a in response["data"]],
            pagination=Pagination(**response["pagination"]),
        )
